doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
+
+Since backpressure is not feasible right off the bat when reading from materialized view as I had
+assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
+doing the [multi part uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
+to S3.
+* A single worker starts a multi part upload with a `CreateMultipartUpload` API call to start
+writing data to a S3 file in the given path.
+* The worker should keep a counter, say `bytes_to_write` to keep track of the aggregate bytes it will write.
+* The worker should also keep some state to track the number of parts in the upload
+and the corresponding etags returned from AWS for the uploaded parts.
+* The worker receives data to be written to S3, adds it to `bytes_to_write` to see if it will exceed
+MAX FILE SIZE.
+* If less than MAX FILE SIZE, the worker assigns an incremental part number starting from 1,
+makes a `UploadPart` call and keeps the returned `ETag` along with the part number in memory.
+* If more than MAX FILE SIZE, the worker makes a `CompleteMultipartUpload` call to finish
+the upload of a single file and provides the part number and etags in the request.
+* Then the worker creates a new multi part request to write to a new S3, resets the
+`bytes_to_write`, clears the part number with etag mapping in memory and uploads the part for
+the new multi part call.
+
+#### S3 file names
+The name of the individual S3 files can follow a pattern like `part-0001.csv`, `part-0002.csv` etc.
+
+## Rollout and Testing
+We should put COPY TO S3 behind a feature flag.
+
+#### Testing During development
+TODO: Check how we do persist tests
+
+#### Testing after code is merged
+- Switch on the feature flag for the staging environment and do an end to end copy to s3
+flow in staging.
+- Also, try out a scenario where we cancel the copy midway
+
+## Possible Future work
+
+### Parquet output format
+As mentioned earlier, there's already user interest for parquet. That can be a follow-up to this
+feature. We'll need a comprehensive type mapping between parquet and materialize types and we
+should do a separate design doc for that. Ideally that can be re-used for `COPY FROM S3` as well
+if we ever want to support that.
+
+### More observability
+Postgres reports the progress of an ongoing `COPY` to a `pg_stat_progress_copy`. We can have a
+similar table to show progress which the user can query to see how far along they are.
+
```

Comment:
Worth also calling out:

* *Partitions:* Since the target use case here is unloading data to an analytical data warehouse downstream, users will probably ask about partitions sooner or later for query performance.

* *Scheduling:* Most users will have an orchestrator or manage custom scripts for executing jobs like these, but certainly some will want to set a copy out schedule in Materialize. We should support this.

Reply:
TBH I am unsure about Scheduling and if it's something we should support in Materialize, but I am adding `Partitions` as potential future work item.